### PR TITLE
memory: do not set_reclaim_hook if cpu_mem_ptr is not set

### PR DESCRIPTION
**Pull request by tchaikov**
in the release build, we use Seastar allocator, but the Seastar application
can still configure `smp_opts.memory_allocator` with `memory_allocator::standard`.
in that case, the `memory::configure()` is not called, hence `cpu_mem_ptr` is not
set.

but in the constructor of `reactor`, it is supposed to setup per-reactor
facilities, notably the ones backed by TLS variables, including the
memory subsystem. but we are not able to access the `smp_opts` anymore
in `reactor::reactor()`. because the memory subsystem is configured in
`smp::configure()`, where is `smp_opts` passed in.

so, in this change

* instead of conditionally calling
  `memory::set_reclaim_hook()` in `reactor::reactor()` based on the
  value of `smp_opts.memory_allocator`, we check for null in
  `memory::set_reclaim_hook()`. fortunately, this is not a hot path,
  we only call `memory::set_reclaim_hook()` when starting the reactor,
  so the extra check does not hurt the performance.
* add a test for exercising template_app, as the smp options is
  a public interface for customizing the behavior of app_template.
  because boot.test considers it an error if a test suite does not
  contains any test, and fails the test with:
  ```
  Test setup error: test tree is empty
  ```
  we cannot disable the test in app-template_test.cc using
  `boost::unit_test::enable_if<>`. neither can we add more
  test cases in this test, because seastar cannot be initialized
  multiple times in a single process. so we have to disable
  this test in the CMake.

### CODE REVIEW

File: tests/unit/app_test.cc
Comment (tchaikov) [inline tests/unit/app_test.cc:34]: i am trying to address this in #2001

File: tests/unit/app_test.cc
Comment (denesb) [inline tests/unit/app_test.cc:51]: Maybe add a trivial check here that the reactor didn't silently bork.

```c++
const auto int64_t canary_magic = 76868676876;
int64_t canary = 0;
int exit_code = app.run(std::size(args), std::data(args), [&] {
    canary = canary_magic;
    return make_ready_future();
});
BOOST_REQUIRE_EQUAL(canary, canary_magic);
```

Comment (tchaikov) [reply tests/unit/app_test.cc]: i tweak the test a little bit so it returns the expected `exit_code` to verify that the app template executes the coroutine for us.

**Review by denesb — APPROVED** [review]

File: tests/unit/app_test.cc
Comment (avikivity) [inline tests/unit/app_test.cc:47]: (that's to be compatible with main() signature)

**Comment by avikivity** [comment]
> The test segfaults in `smp::configure()` with the dpdk build consistently. But I cannot reproduce the failure locally with the same setting so far.

It's possible that the standard allocator and dpdk are incompatible with each other.

File: tests/unit/app_test.cc
Comment (tchaikov) [reply tests/unit/app_test.cc:56]: thank you Avi.

**Comment by tchaikov** [comment]
> > The test segfaults in `smp::configure()` with the dpdk build consistently. But I cannot reproduce the failure locally with the same setting so far.
> 
> It's possible that the standard allocator and dpdk are incompatible with each other.

i created #2003 to track this. tried to debug this, but i cannot figure out why. disabling the test for the dpdk build at this moment.

**Comment by tchaikov** [comment]
v2:

- disable the newly added with dpdk build due to #2003

**Comment by tchaikov** [comment]
in order to ~address~ workaround the test failure, i skip the test when seastar is built with dpdk enabled, @denesb could you take another look?

**Review by denesb — APPROVED** [review]

**Comment by tchaikov** [comment]
@scylladb/seastar-maint hello maintainers, could you help merge this change?

File: tests/unit/CMakeLists.txt
Comment (avikivity) [inline tests/unit/CMakeLists.txt:415]: It's better to do this in the test case itself, in case we add more tests to app-template_test.cc which can work with dpdk.

Comment (avikivity) [reply tests/unit/CMakeLists.txt:415]: Plus, less cmake = better

Comment (tchaikov) [reply tests/unit/CMakeLists.txt:415]: @avikivity i explained the rationales in the commit message, why i cannot disable it in the test. could you please take a look? see https://github.com/scylladb/seastar/pull/2000/commits/b1ddb7ff94be7fb6ee282f14ff3f8375370e5e77

Comment (tchaikov) [reply tests/unit/CMakeLists.txt:415]: i will be able to revert the cmake change once #2001 gets merged.

Comment (tchaikov) [reply tests/unit/CMakeLists.txt:415]: this test now conflicts with the latest revision of #2001. i will rebase it once #2001 is merged.

**Pull request merged**
